crates/bevy_ecs/src/schedule/state.rs

```diff
@@ -181,56 +227,781 @@ impl<S: States> NextState<S> {
+
+                    world.send_event(StateTransitionEvent {
+                        before: None,
+                        after: Some(entered.clone()),
+                    });
+
+                    let mut state_transition_schedules =
+                        world.get_resource_or_insert_with(StateTransitionSchedules::default);
+                    state_transition_schedules
+                        .enter_schedules
+                        .entry(S::DEPENDENCY_DEPTH)
+                        .or_default()
+                        .insert(OnEnter(entered.clone()).intern());
+                }
+            };
+        }
+        None => {
+            // We first remove the [`State<S>`] resource, and if one existed we compute dependant states, send a transition event and run the `OnExit` schedule.
+            if let Some(resource) = world.remove_resource::<State<S>>() {
+                world
+                    .try_run_schedule(ComputeDependantStates::<S>::default())
+                    .ok();
+
+                world.send_event(StateTransitionEvent {
+                    before: Some(resource.get().clone()),
+                    after: None,
+                });
+
+                let mut state_transition_schedules =
+                    world.get_resource_or_insert_with(StateTransitionSchedules::default);
+                state_transition_schedules
+                    .exit_schedules
+                    .entry(S::DEPENDENCY_DEPTH)
+                    .or_default()
+                    .insert(OnExit(resource.0).intern());
+            }
+        }
+    }
+}
+
+fn prepare_state_transitions(world: &mut World) {
+    world.insert_resource(StateTransitionSchedules::default());
+}
+
+fn execute_state_transition_schedules(world: &mut World) {
+    if let Some(schedules) = world.remove_resource::<StateTransitionSchedules>() {
+        for (_, schedules) in schedules.exit_schedules {
+            for schedule in schedules {
+                let _ = world.try_run_schedule(schedule);
+            }
+        }
+
+        for (_, schedules) in schedules.transition_schedules {
+            for schedule in schedules {
+                let _ = world.try_run_schedule(schedule);
+            }
+        }
+
+        for (_, schedules) in schedules.enter_schedules {
+            for schedule in schedules {
+                let _ = world.try_run_schedule(schedule);
+            }
+        }
+    }
+}
+
+fn execute_state_transitions(world: &mut World) {
+    prepare_state_transitions(world);
+    let _ = world.try_run_schedule(ManualStateTransitions);
+    while let Some((_, compute)) = world
+        .get_resource_mut::<StateTransitionSchedules>()
+        .and_then(|mut schedules| schedules.dependant_schedules.pop_first())
+    {
+        for schedule in compute {
+            let _ = world.try_run_schedule(schedule);
+        }
+    }
+    execute_state_transition_schedules(world);
+}
+
+#[derive(Resource)]
+struct StateTransitionsHaveBeenSetup;
```

Comment:
Is this needed? Can't we just check if the `StateTransition` schedule is added?

Reply:
I don't know why I didn't think of that XD
done